src/main/java/com/auth0/RequestProcessor.java

```diff
@@ -166,40 +186,78 @@ private Tokens getVerifiedTokens(HttpServletRequest req, Tokens frontChannelToke
      *
-     * @param req the request
+     * @param request the request
      * @return a new instance of Tokens wrapping the values present in the request parameters.
      */
-    private Tokens getFrontChannelTokens(HttpServletRequest req) {
-        Long expiresIn = req.getParameter(KEY_EXPIRES_IN) == null ? null : Long.parseLong(req.getParameter(KEY_EXPIRES_IN));
-        return new Tokens(req.getParameter(KEY_ACCESS_TOKEN), req.getParameter(KEY_ID_TOKEN), null, req.getParameter(KEY_TOKEN_TYPE), expiresIn);
+    private Tokens getFrontChannelTokens(HttpServletRequest request) {
+        Long expiresIn = request.getParameter(KEY_EXPIRES_IN) == null ? null : Long.parseLong(request.getParameter(KEY_EXPIRES_IN));
+        return new Tokens(request.getParameter(KEY_ACCESS_TOKEN), request.getParameter(KEY_ID_TOKEN), null, request.getParameter(KEY_TOKEN_TYPE), expiresIn);
     }
 
     /**
      * Checks for the presence of an error in the request parameters
      *
-     * @param req the request
+     * @param request the request
      * @throws InvalidRequestException if the request contains an error
      */
-    private void assertNoError(HttpServletRequest req) throws InvalidRequestException {
-        String error = req.getParameter(KEY_ERROR);
+    private void assertNoError(HttpServletRequest request) throws InvalidRequestException {
+        String error = request.getParameter(KEY_ERROR);
         if (error != null) {
-            String errorDescription = req.getParameter(KEY_ERROR_DESCRIPTION);
+            String errorDescription = request.getParameter(KEY_ERROR_DESCRIPTION);
             throw new InvalidRequestException(error, errorDescription);
         }
     }
 
     /**
-     * Checks whether the state persisted in the session matches the state value received in the request parameters.
+     * Checks whether the state received in the request parameters is the same as the one in the state cookie or session
+     * for this request.
      *
-     * @param req the request
+     * @param request the request
      * @throws InvalidRequestException if the request contains a different state from the expected one
      */
-    private void assertValidState(HttpServletRequest req) throws InvalidRequestException {
-        String stateFromRequest = req.getParameter(KEY_STATE);
-        boolean valid = RandomStorage.checkSessionState(req, stateFromRequest);
+    private void assertValidState(HttpServletRequest request, HttpServletResponse response) throws InvalidRequestException {
+        String stateFromRequest = request.getParameter(KEY_STATE);
+
+        // If response is null, check the Session.
+        // This can happen when the deprecated handle method that only takes the request parameter is called
+        if (response == null) {
+            checkSessionState(request, stateFromRequest);
+            return;
+        }
+
+        String actualState = TransientCookieStore.getState(request, response, legacySameSiteCookie);
```

Comment:
evaluate renaming this prop

Reply:
Also evaluate some pattern like this one, to move the fallback logic into a diff class.

```java
interface Storage {
  getNonce()
  getState()
  getValue(String keyName)
}
```

//CookieStorage implements Storage
```java

class FallbackStorage implements Storage{
   private Storage prevInstance;
   FallbackStorage(Storage instance){
     //constructor
     this.prevInstance = instance
   }

   getNonce(){
    String nonce = prevInstance.getNonce()
       if (nonce==null){
          //TODO get from request
       }
    }
}
```

```java
Storage buildStorage() {
  Storage storage  = new CookieStorage(req, res)
  storage.setFlag(SSflag)
  if (shouldUseFallback?){
    val storage = new FallbackStorage(storage)
  }
  return storage
}
```

Reply:
Renamed `actualState` to `cookieState`. Will consider Storage refactoring and whether it should be in a separate PR.

Reply:
I'd like to consider such a refactoring separate from this PR, and once we get this and the deprecations in. Since it's internal we can refactor later, and consider any impacts it might have when moving away from Session storage entirely.

Reply:
I agree!